azure-iot-device/azure/iot/device/common/pipeline/operation_flow.py

```diff
@@ -15,124 +15,145 @@
 logger = logging.getLogger(__name__)
 
 
-@pipeline_thread.runs_on_pipeline_thread
-def delegate_to_different_op(stage, original_op, new_op):
-    """
-    Continue an operation using a new operation.  This means that the new operation
-    will be passed down the pipeline (starting at the next stage). When that new
-    operation completes, the original operation will also complete.  In this way,
-    a stage can accept one type of operation and, effectively, change that operation
-    into a different type of operation before passing it to the next stage.
-
-    This is useful when a generic operation (such as "enable feature") needs to be
-    converted into a more specific operation (such as "subscribe to mqtt topic").
-    In that case, a stage's _execute_op function would call this function passing in
-    the original "enable feature" op and the new "subscribe to mqtt topic"
-    op.  This function will pass the "subscribe" down. When the "subscribe" op
-    is completed, this function will cause the original op to complete.
-
-    This function is only really useful if there is no data returned in the
-    new_op that that needs to be copied back into the original_op before
-    completing it.  If data needs to be copied this way, some other method needs
-    to be used.  (or a "copy data back" function needs to be added to this function
-    as an optional parameter.)
-
-    :param PipelineStage stage: stage to delegate the operation to
-    :param PipelineOperation original_op: Operation that is being continued using a
-      different op.  This is most likely the operation that is currently being handled
-      by the stage.  This operation is not actually continued, in that it is not
-      actually passed down the pipeline.  Instead, the original_op operation is
-      effectively paused while we wait for the new_op operation to complete.  When
-      the new_op operation completes, the original_op operation will also be completed.
-    :param PipelineOperation new_op: Operation that is being passed down the pipeline
-      to effectively continue the work represented by original_op.  This is most likely
-      a different type of operation that is able to accomplish the intention of the
-      original_op in a way that is more specific than the original_op.
-    """
-
-    logger.debug("{}({}): continuing with {} op".format(stage.name, original_op.name, new_op.name))
+class PipelineFlow(object):
```

Comment:
Not sure this is an ideal name - not really a noun.

Reply:
Noun, schmoun.  Flow is a noun, as in "this is how the flow through the pipeline happens".

---
In reply to: [333724482](https://github.com/Azure/azure-iot-sdk-python/pull/347#discussion_r333724482) [](ancestors = 333724482)